### PR TITLE
Vanilla - added generic PSR-compatible middleware

### PR DESCRIPTION
**Pull request by itsgoingd**
- Adds a generic PSR-compatible middleware to the vanilla integration.
- For use with applications and frameworks implementing [PSR-15](https://www.php-fig.org/psr/psr-15/) (and by extension [PSR-7](https://www.php-fig.org/psr/psr-7) and [PSR-17](https://www.php-fig.org/psr/psr-17)), eg. [Slim](https://www.slimframework.com), [Mezzio](https://docs.mezzio.dev), etc.
- Also improves the overall vanilla integration - using PSR request instead of globals if specified, allowing to set or auto-detect a PSR response factory.
- To allow the vanilla integration to generate responses a PSR-compatible response factory has to be provided, or the `php-http/discovery` has to be installed for zero-configuration use.
- This will replace the Slim-specific integration in a future major version.
- Based on an earlier [PR](https://github.com/itsgoingd/clockwork/pull/680) by @UlrichEckhardt and a [PR](https://github.com/itsgoingd/clockwork/pull/675) by @DominicDetta

Example use in a Slim application:

```php
use Clockwork\Support\Vanilla\ClockworkMiddleware;
$app->add(ClockworkMiddleware::init()->withResponseFactory($app->getResponseFactory()));

// Run App & Emit Response
$response = $app->handle($request);
```

With the `php-http/discovery` package installed, the response factory will be auto-detected (if one is installed), making the initialization as simple as:

```php
use Clockwork\Support\Vanilla\ClockworkMiddleware;
$app->add(ClockworkMiddleware::init());
```

You can also instantiate the middleware with a custom Clockwork instance (useful if you want to register the instance with a DI container for example) and disable routing if you want to set it up on your own:

```php
use Clockwork\Support\Vanilla\Clockwork;
use Clockwork\Support\Vanilla\ClockworkMiddleware;
$clockwork = new Clockwork([ ... ]);
$app->add((new ClockworkMiddleware($clockwork))->withoutRouting());

$app->get('/__clockwork/{request:.+}', function ($request, $response) use ($clockwork) {
    return $clockwork->usePsrMessage($request, $response)->handleMetadata();
});
```

Note, because the Slim skeleton project registers the routing middleware as the last (outermost) middleware, you will need to add Clockwork middleware in `public/index.php` just before the `$app->handle()` call, instead of `app/middleware.php`.

### CODE REVIEW

**Comment by itsgoingd** [comment]
@UlrichEckhardt @DominicDetta Please check this PR out and let me know what you think about my take on this feature.

File: Clockwork/Support/Vanilla/ClockworkMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
+<?php namespace Clockwork\Support\Vanilla;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+// Generic Clockwork middleware fro use with PSR-compatible applications
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Vanilla/ClockworkMiddleware.php:9]: Typo: "fro" -> "for"

File: Clockwork/Support/Vanilla/ClockworkMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	public function withoutRouting()
+	{
+		$this->handleRouting = false;
+		return $this;
+	}
+	
+	// Process the middleware
+	public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) :ResponseInterface
+	{
+		$request->withAttribute('clockwork', $this->clockwork);
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Vanilla/ClockworkMiddleware.php:51]: PSR requests are immutable, so this doesn't do what you intend. It returns a new request with the attribute attached.

File: Clockwork/Support/Vanilla/ClockworkMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		
+		$this->clockwork->event('Controller')->begin();
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Vanilla/ClockworkMiddleware.php:53]: This lacks an according call to `event(..)->end()`.

File: Clockwork/Support/Vanilla/Clockwork.php [hunk 4]
```diff
@@ -397,10 +402,10 @@ protected function configureSerializer()
...
 	{
 		$this->clockwork->shouldCollect([
-			'onDemand'        => $this->config['requests']['on_demand'],
-			'sample'          => $this->config['requests']['sample'],
-			'except'          => $this->config['requests']['except'],
-			'only'            => $this->config['requests']['only'],
+			'onDemand'       => $this->config['requests']['on_demand'],
+			'sample'         => $this->config['requests']['sample'],
+			'except'         => $this->config['requests']['except'],
+			'only'           => $this->config['requests']['only'],
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Vanilla/Clockwork.php:408]: Is this change intentional?

File: Clockwork/Support/Vanilla/Clockwork.php [hunk 5]
```diff
@@ -456,14 +461,29 @@ protected function response($data = null, $status = null, $json = true)
...
+
+		return $this->incomingRequest = new IncomingRequest([
+			'method'  => $this->psrRequest ? $this->psrRequest->getMethod() : $_SERVER['REQUEST_METHOD'],
+			'uri'     => $this->psrRequest ? $this->psrRequest->getUri()->getPath() : $_SERVER['REQUEST_URI'],
+			'headers' => $this->psrRequest
+				? array_map(function ($values) { return implode(', ', $values); }, $this->psrRequest->getHeaders())
+				: $_SERVER,
+			'input'   => $this->psrRequest
+				? array_merge($this->psrRequest->getQueryParams(), (array) $this->psrRequest->getParsedBody())
+				: array_merge($_GET, $_POST, (array) json_decode(file_get_contents('php://input'), true))
```
Comment (UlrichEckhardt) [inline Clockwork/Support/Vanilla/Clockwork.php:474]: I'm not sure whether the `'input'` has the same content with the two implementations. 

That said, I believe this could become a bit easier to read if it checked `$this->psrRequest` only once:

```php
return $this->incomingRequest = new IncomingRequest(
    $this->psrRequest ? [
        'method' => $this->psrRequest->getMethod(),
        'uri'    => $this->psrRequest->getUri()->getPath(),
        ..
    ] : [
        'method' => $_SERVER['REQUEST_METHOD'],
        'uri'    => $_SERVER['REQUEST_URI'],
        ..
    ]
);

**Review by UlrichEckhardt — COMMENTED** [review]
Just looked at the code a bit, so I haven't tried it yet, but that's now on my list for today. Here are a few remarks up front already.

File: Clockwork/Support/Vanilla/ClockworkMiddleware.php (hunk 3 above)
Comment (itsgoingd) [reply Clockwork/Support/Vanilla/ClockworkMiddleware.php:53]: The running events implicitly end when the Clockwork metadata is resolved (this is just a very rough approximation of what the "controller" runtime could be, as we have no idea what middleware is running before/after Clockwork.

File: Clockwork/Support/Vanilla/Clockwork.php (hunk 4 above)
Comment (itsgoingd) [reply Clockwork/Support/Vanilla/Clockwork.php:408]: Yeah, just fixing some white-space. :)

File: Clockwork/Support/Vanilla/Clockwork.php (hunk 5 above)
Comment (itsgoingd) [reply Clockwork/Support/Vanilla/Clockwork.php]: Fixed, now it should work even for metadata update requests with JSON bodies. Also split into separate methods to look more clean.

**Comment by itsgoingd** [comment]
Thanks for the review, everything should be fixed now.

**Comment by itsgoingd** [comment]
An example Slim project is now also available here - https://github.com/underground-works/clockwork-examples/tree/master/Integrations/Slim#readme

File: Clockwork/Support/Vanilla/Clockwork.php (hunk 4 above)
Comment (UlrichEckhardt) [reply Clockwork/Support/Vanilla/Clockwork.php:408]: But this unfixed it, the last key `'exceptPreflight'` is now misaligned with the rest.

**Comment by UlrichEckhardt** [comment]
Just updated my demo repository:

- Branch [feature/vanilla-psr-middleware](https://github.com/UlrichEckhardt/clockwork-slim-demo/tree/feature/vanilla-psr-middleware) -- this one's based on the code from this PR.
- Branch [feature/psr-17-middleware](https://github.com/UlrichEckhardt/clockwork-slim-demo/tree/feature/psr-17-middleware) -- This one's on my own implementation, just for comparison.

The reason I'm mentioning this is that both of these have PHPUnit tests included, which check whether the demo API, the Clockwork API and the Clockwork web UI works. The tests there differ also a bit because of my uncertainty about where the web UI is supposed to be located URL-wise.

Cheers! :)

File: Clockwork/Support/Vanilla/Clockwork.php (hunk 4 above)
Comment (itsgoingd) [reply Clockwork/Support/Vanilla/Clockwork.php:408]: Haha, the code editor I am using now, uses ligatures by default (fl in flight). Not a great idea in a code editor.

Screenshot 2024-03-05 at 18 13 14

**Comment by itsgoingd** [comment]
Thanks for the feedback! Yes, I assume in the examples, that the enable behavior is already changed. That will be done in a separate PR though.

Oops, the default setting for the web UI assets public path was one level off, so they were not actually copied to the public dir as they were supposed to. This should be resolved in the latest commit.

So let me try to explain how the paths work (and good thing you've mentioned this, because I've realised we are missing a bit of the implementation here):
- The default api path is `/__clockwork`. This is typically not configurable (eg. in the Laravel integration) as it is unlikely to clash with any existing routes and having a well-known default makes things easier for the client apps. We do allow to customize this in the Vanilla integration as an exception, mainly just to support the most primitive apps, where the app is literally just a bunch of `.php` files in a directory and you can't even route path like `/__clockwork/whatever`.
- The default web UI path is `/clockwork`. This is configurable in all integrations, as it is more likely to clash with other existing routes. Now in Laravel, there is a route catching all requests for the web UI assets and serving them manually. In the Vanilla integration, to keep things simple, the assets are copied directly to the public directory on first use, and the asset handling is left to the web server.
- Now when the app doesn't run in the root of the domain, but has a base path, let's say `/app`, the paths should be `/app/__clockwork` and `/app/clockwork`. This is obvious, as requests without the `/app` base might not be even routed to the app. An extra header also has to be set, to tell the browser extension what the base path is. In Laravel we auto-detect this base path, but I'm not sure how to do this in a generic way with the PSR-7 request. I guess we might need to require a manually configured path in that case.

I kind of agree the Vanilla helper class is starting to get a little bit complicated and might be a good candidate for a rework at some point. I'm currently assuming a 5.x release for this feature, so I didn't want to do any big changes at this time.

I especially don't like that part, where the response factory gets randomly passed as a third argument to the `usePsrMessage()` method. I might still move that part to the middleware and pass in an empty response ready to be used.

**Comment by UlrichEckhardt** [comment]
Heya! 

Let's merge this, and then make it better. It definitely adds value already, so there is no point in holding it back.

Concerning the things to improve, the web UI isn't working out of the box for me yet. Key to this is actually documented: "Path where to install the Web UI assets, should be publicly accessible". For me, this is not the case since I route every request through my Slim stack. Yes, for e.g. Nginx, you can tell it to just serve the file if it exists and is not a `*.php` and that should be the default if you care for performance. However, it doesn't have to be the case and since I don't serve a website (with some static content) but an API, I'm not be prepared for serving of static files.

Let's get this moving!

Uli

**Comment by UlrichEckhardt** [comment]
That last changes looks fine, too. Let's merge this!
